File: sdk/typescript/src/adapters/llamaindex.ts
/**
 * LlamaIndex adapter for ADP v0.3.0.
 *
 * Converts between ADP manifests and LlamaIndex configurations.
 */

import { AdapterBase } from "./base";
import { AdapterRegistry } from "./registry";
import { ADP } from "../adp";

export class LlamaIndexAdapter extends AdapterBase {
  readonly frameworkId = "llamaindex";

  constructor() {
    super();
    AdapterRegistry.register(LlamaIndexAdapter);
  }

  export(manifest: ADP): Record<string, unknown> {
    const data = manifest as Record<string, unknown>;

    const tools = (data.tools as Record<string, unknown>) ?? {};
    const memory = (data.memory as Record<string, unknown>) ?? {};

    // Map tools to LlamaIndex tools
    const indexTools: Array<Record<string, unknown>> = [];

    const toolListKeys: string[] = ["mcp_servers", "http_apis", "sql_functions"];
    for (const toolListKey of toolListKeys) {
      const toolList = (tools[toolListKey] as Array<Record<string, unknown>>) ?? [];
      for (const tool of toolList) {
        indexTools.push({
          name: tool.id as string,
          description: (tool.description as string) ?? "",
          function: tool.id as string,
        });
      }
    }

    // Map memory stores to LlamaIndex memory
    const memoryConfig: Record<string, unknown> = {};
    if (memory && typeof memory === "object") {
      const stores = (memory.stores as Array<Record<string, unknown>>) ?? [];
      for (const store of stores) {
        const storeType = (store.type as string) ?? "semantic";
        if (storeType === "semantic") {
          memoryConfig.vector_store = {
            provider: store.provider,
            index: store.index,
          };
        } else if (storeType === "episodic") {
          memoryConfig.chat_memory = {
            provider: store.provider,
          };
        }
      }
    }

    // Build QueryEngine configuration
    const queryEngine: Record<string, unknown> = {
      tools: indexTools,
      memory: memoryConfig,
    };

    // Apply adapter hints
    const adapterHints = (
      (data.runtime as Record<string, unknown>)?.adapter_hints as Record<string, unknown>
    )?.llamaindex as Record<string, unknown> | undefined;

    if (adapterHints?.embedder_config !== undefined) {
      queryEngine.embedder = adapterHints.embedder_config;
    }

    return queryEngine;
  }

  importFrom(config: Record<string, unknown>): ADP {
    const tools = (config.tools as Array<Record<string, unknown>>) ?? [];
    const memoryConfig = (config.memory as Record<string, unknown>) ?? {};
    const embedder = config.embedder as Record<string, unknown> | undefined;

    // Convert tools to ADP tools
    const adpTools: Array<Record<string, unknown>> = [];
    for (const tool of tools) {
      adpTools.push({
        id: tool.name as string,
        description: (tool.description as string) ?? "",
        base_url: (tool.base_url as string) ?? "",
      });
    }

    // Convert memory to ADP memory
    const adpMemory: Record<string, unknown> = {};
    if (memoryConfig.vector_store !== undefined) {
      const vs = memoryConfig.vector_store as Record<string, unknown>;
      adpMemory.stores = [{
        id: "vector_store",
        type: "semantic",
        provider: vs.provider,
        index: vs.index,
      }];
    }
    if (memoryConfig.chat_memory !== undefined) {
      const cm = memoryConfig.chat_memory as Record<string, unknown>;
      if (!adpMemory.stores) {
        adpMemory.stores = [];
      }
      (adpMemory.stores as Array<Record<string, unknown>>).push({
        id: "chat_memory",
        type: "episodic",
        provider: cm.provider,
      });
    }

    // Extract adapter hints
    const adapterHints: Record<string, unknown> = {};
    if (embedder) {
      adapterHints.embedder_config = embedder;
    }

    return {
      adp_version: "0.3.0",
      id: (config.id as string) ?? "imported-from-llamaindex",
      runtime: {
        execution: [
          {
            backend: "python",
            id: "llamaindex",
            module: "llama_index",
          },
        ],
      },
      flow: {
        id: "imported-flow",
        graph: {
          nodes: [
            {
              id: "query",
              kind: "llm",
            },
          ],
          edges: [],
          start_nodes: ["query"],
          end_nodes: ["query"],
        },
        extensions: adapterHints && Object.keys(adapterHints).length > 0 ? { llamaindex: adapterHints } : undefined,
      },
      tools: adpTools.length > 0 ? { http_apis: adpTools } : undefined,
      memory: adpMemory && Object.keys(adpMemory).length > 0 ? adpMemory : undefined,
      evaluation: { suites: [] },
      extensions: {
        source_framework: "llamaindex",
      },
    } as unknown as ADP;
  }

  roundtripFidelity(): Record<string, string> {
    return {
      ...super.roundtripFidelity(),
      "memory.stores": "faithful",
      "tools": "faithful",
    };
  }
}

File: sdk/python/adp_sdk/adapters/llamaindex.py
"""LlamaIndex adapter for ADP v0.3.0.

Converts between ADP manifests and LlamaIndex configurations.
"""

from typing import Any

from .base import AdapterBase
from .registry import AdapterRegistry
from ..adp_model import ADP


class LlamaIndexAdapter(AdapterBase):
    """Adapter for LlamaIndex framework.

    Converts ADP tools and memory to LlamaIndex QueryEngine configurations.
    """

    framework_id = "llamaindex"

    def __init__(self) -> None:
        AdapterRegistry.register(self.__class__)

    def export(self, manifest: ADP) -> dict[str, Any]:
        """Export ADP manifest to LlamaIndex config."""
        data = manifest.model_dump(by_alias=True, exclude_none=True)

        tools = data.get("tools", {})
        memory = data.get("memory", {})

        # Map tools to LlamaIndex tools
        index_tools: list[dict[str, Any]] = []

        for tool_list_key in ("mcp_servers", "http_apis", "sql_functions"):
            for tool in tools.get(tool_list_key, []):
                index_tools.append({
                    "name": tool.get("id"),
                    "description": tool.get("description", ""),
                    "function": tool.get("id"),
                })

        # Map memory stores to LlamaIndex memory
        memory_config: dict[str, Any] = {}
        if isinstance(memory, dict):
            stores = memory.get("stores", [])
            for store in stores:
                store_type = store.get("type", "semantic")
                if store_type == "semantic":
                    memory_config["vector_store"] = {
                        "provider": store.get("provider"),
                        "index": store.get("index"),
                    }
                elif store_type == "episodic":
                    memory_config["chat_memory"] = {
                        "provider": store.get("provider"),
                    }

        # Build QueryEngine configuration
        query_engine: dict[str, Any] = {
            "tools": index_tools,
            "memory": memory_config,
        }

        # Apply adapter hints
        adapter_hints = (data.get("runtime", {}).get("adapter_hints", {})
                         .get("llamaindex", {}))

        if embedder_config := adapter_hints.get("embedder_config"):
            query_engine["embedder"] = embedder_config

        return query_engine

    def import_from(self, config: dict[str, Any]) -> ADP:
        """Import LlamaIndex config to ADP manifest."""
        tools = config.get("tools", [])
        memory_config = config.get("memory", {})
        embedder = config.get("embedder")

        # Convert tools to ADP tools
        adp_tools = []
        for tool in tools:
            adp_tools.append({
                "id": tool.get("name", ""),
                "description": tool.get("description", ""),
                "base_url": tool.get("base_url", ""),
            })

        # Convert memory to ADP memory
        adp_memory: dict[str, Any] = {}
        if vector_store := memory_config.get("vector_store"):
            adp_memory["stores"] = [{
                "id": "vector_store",
                "type": "semantic",
                "provider": vector_store.get("provider"),
                "index": vector_store.get("index"),
            }]
        if chat_memory := memory_config.get("chat_memory"):
            if "stores" not in adp_memory:
                adp_memory["stores"] = []
            adp_memory["stores"].append({
                "id": "chat_memory",
                "type": "episodic",
                "provider": chat_memory.get("provider"),
            })

        # Extract adapter hints
        adapter_hints = {}
        if embedder:
            adapter_hints["embedder_config"] = embedder

        from ..adp_model import (
            RuntimeEntry,
            RuntimeModel,
            GraphModel,
            FlowModel,
            ToolsModel,
            HTTPAPIModel,
        )

        http_apis = [
            HTTPAPIModel(
                id=t.get("name", ""),
                description=t.get("description", ""),
                base_url=t.get("base_url", ""),
            )
            for t in tools
        ] if tools else None

        return ADP(
            adp_version="0.3.0",
            id=config.get("id", "imported-from-llamaindex"),
            runtime=RuntimeModel(execution=[RuntimeEntry(
                backend="python",
                id="llamaindex",
                module="llama_index",
            )]),
            flow=FlowModel(
                id="imported-flow",
                graph=GraphModel(
                    nodes=[{
                        "id": "query",
                        "kind": "llm",
                    }],
                    edges=[],
                    start_nodes=["query"],
                    end_nodes=["query"],
                ),
                extensions={"llamaindex": adapter_hints} if adapter_hints else None,
            ),
            tools=ToolsModel(http_apis=http_apis) if http_apis else None,
            memory=adp_memory if adp_memory else None,
            extensions={"source_framework": "llamaindex"},
        )

    def roundtrip_fidelity(self) -> dict[str, str]:
        return {
            **super().roundtrip_fidelity(),
            "memory.stores": "faithful",
            "tools": "faithful",
        }
